File: 2023/src/main/scala/day01.scala
package com.markkovari.adventofcode.day01

import scala.collection.immutable.HashMap
import scala.io.Source

val numbersAsStringsAndValues =
  HashMap(
    "one" -> 1,
    "two" -> 2,
    "three" -> 3,
    "four" -> 4,
    "five" -> 5,
    "six" -> 6,
    "seven" -> 7,
    "eight" -> 8,
    "nine" -> 9
  )

def stringStartsStringifiedDigit(text: String): Option[Int] =
  numbersAsStringsAndValues
    .find { case (key, _) => text.startsWith(key) }
    .map { case (_, value) => value }

def getFirstAndLastMultipliedTen(list: List[Int]): Int =
  list match {
    case a :: Nil       => a * 10 + a
    case a :: b :: Nil  => a * 10 + b
    case a :: b :: tail => a * 10 + tail.last
    case _              => 0
  }

def getMixedUpDigits(of: String): List[Int] = of match {
  case "" => List()
  case other =>
    stringStartsStringifiedDigit(other) match {
      case Some(value) => value :: getMixedUpDigits(other.splitAt(1)._2)
      case None =>
        val (head, tail) = other.splitAt(1)
        head.toIntOption match {
          case None        => getMixedUpDigits(tail)
          case Some(value) => value :: getMixedUpDigits(tail)
        }
    }
}

def getDigits(of: String): List[Int] = of match {
  case "" => List()
  case a =>
    val (head, tail) = a.splitAt(1)
    head.toIntOption match {
      case None        => getDigits(tail)
      case Some(value) => value :: getDigits(tail)
    }
}

File: 2023/rust/src/day01.rs
use anyhow::Result;
use aoc_rust_common::Solution;

pub struct Day01;

impl Solution for Day01 {
    fn year(&self) -> u32 {
        2023
    }
    fn day(&self) -> u32 {
        1
    }

    fn part1(&self, input: &str) -> Result<String> {
        let result: u32 = input
            .lines()
            .map(|line| {
                let digits: Vec<u32> = line.chars().filter_map(|c| c.to_digit(10)).collect();
                digits.first().unwrap() * 10 + digits.last().unwrap()
            })
            .sum();
        Ok(result.to_string())
    }

    fn part2(&self, input: &str) -> Result<String> {
        let result: u32 = input
            .lines()
            .map(|line| {
                let digits = get_mixed_up_digits(line);
                digits.first().unwrap() * 10 + digits.last().unwrap()
            })
            .sum();
        Ok(result.to_string())
    }
}

fn get_mixed_up_digits(line: &str) -> Vec<u32> {
    let mut digits = Vec::new();
    for i in 0..line.len() {
        let slice = &line[i..];
        if slice.starts_with("one") {
            digits.push(1);
        } else if slice.starts_with("two") {
            digits.push(2);
        } else if slice.starts_with("three") {
            digits.push(3);
        } else if slice.starts_with("four") {
            digits.push(4);
        } else if slice.starts_with("five") {
            digits.push(5);
        } else if slice.starts_with("six") {
            digits.push(6);
        } else if slice.starts_with("seven") {
            digits.push(7);
        } else if slice.starts_with("eight") {
            digits.push(8);
        } else if slice.starts_with("nine") {
            digits.push(9);
        } else if let Some(d) = slice.chars().next().unwrap().to_digit(10) {
            digits.push(d);
        }
    }
    digits
}

aoc_rust_common::aoc_test!(Day01, "54573", "54591");
